### PR TITLE
[DAC] ExcludeMemberProposal

### PR DESCRIPTION
**Pull request by evlekht**
## Why this should be merged
- Adds exclude member proposal.
- Extends proposals and finishProposalsTx logic to allow successful proposals unbond additional utxos.
- Extracts proposalState types from txs codec to their own dac codec
- Some general renaming and cleanup over dac

### Exclude member proposal
PR adds new proposal type and corresponding proposals executor and verifier implementations.

If successful, exclude member proposal will do:
- unset consortium member address state bit.
- remove registered node short-link, if any
- suspend active validator
- remove pending validator and return its bond
### BondTxIDsGetter
Because exclude member proposal is capable of unbonding bond other than proposals, finishProposalsTx was extended to use another proposals visitor: BondTxIDsGetter. This visitor is called for successful proposals and will return any additional bond tx ids that should be unlocked along with proposal bonds.

And because finishProposalsTx is system tx that is built by node, txs/builder pkg has to import dac BondTxIDsGetter implementation. In order to avoid cycle dependencies, dac visitors implementations was extracted from txs/executor to their own txs/dac pkg.
### Codec
The reason for creating dac codec for proposalState type is that this types actually not used in transaction bodies, so there is no need for this types to pollute txs codec.
### Commits
PR consists of 3 commits:
- ExcludeMemberProposal and all corresponding implementations including BondTxIDsGetter
- General DAC cleanup
- Codec changes
## How this was tested
unit-tests, integration tests
## Co-authors:
@knikos great help with bondTxIDsGetter improvement

### CODE REVIEW

File: vms/platformvm/txs/executor/camino_helpers_test.go [hunk 1]
```diff
@@ -736,6 +736,7 @@ func expectVerifyUnlockDeposit(
 	expectGetMultisigAliases(t, s, addrs, aliases)
 }
 
+// TODO @evlekht seems, that [addrs] actually not affecting anything and could be omitted
```
Comment (knikos) [inline vms/platformvm/txs/executor/camino_helpers_test.go:739]: is this something you want deal in the future and not now?

File: vms/platformvm/txs/dac/camino_dac.go
Comment (knikos) [inline vms/platformvm/txs/dac/camino_dac.go:1]: i think eventually that it makes sense to extract dac related stuff to a separate folder but this `dac` folder should actually go under executor right?

File: vms/platformvm/txs/dac/camino_dac.go [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+
+// We should always mind possible proposals conflict, when implementing proposal execution logic.
+// Because when proposal is semantically verified, state doesn't know about changes
+// that already existing proposals will bring into state on their execution.
+// And proposal execution is a system tx, so it should always succeed.
+type proposalBondTxIDsGetter struct {
+	state state.Chain
+}
+
+func ProposalVerifier(state state.Chain, fx fx.Fx, signedTx *txs.Tx, tx *txs.AddProposalTx) *proposalVerifier {
```
Comment (knikos) [inline vms/platformvm/txs/dac/camino_dac.go]: i find it confusing that this function does not start with the prefix 'New..'
that applies also to ProposalExecutor and ProposalBondTxIDsGetter functions.
Is there a reason you chose to omit it?
Also, how come don't you define in the function's signature the Interface type instead of the struct pointer?

File: vms/platformvm/camino_helpers_test.go [hunk 3]
```diff
@@ -115,7 +115,7 @@ func newCaminoVM(genesisConfig api.Camino, genesisUTXOs []api.UTXO, startTime *t
 	return vm
 }
 
-func defaultCaminoConfig(postBanff bool) config.Config {
+func defaultCaminoConfig(postBanff bool) config.Config { //nolint:unparam
```
Comment (knikos) [inline vms/platformvm/camino_helpers_test.go:118]: why is that necessary?

File: vms/platformvm/dac/camino_proposal.go [hunk 4]
```diff
@@ -62,6 +70,8 @@ type ProposalState interface {
 	IsSuccessful() bool // should be called only for finished proposals
 	Outcome() any       // should be called only for finished successful proposals
 	Visit(ExecutorVisitor) error
+	// Visits getter and returns additional lock tx ids, that should be unbonded when this proposal is successfully finished.
+	GetBondTxIDs(BondTxIDsGetter) ([]ids.ID, error)
```
Comment (knikos) [inline vms/platformvm/dac/camino_proposal.go]: tbh i would rename GetBondTXIDs to `Invoke` or sthg. Cause I see some repetition GetBondTxID and BondTxIDsGetter

File: vms/platformvm/dac/camino_exclude_member_proposal_test.go [hunk 5]
```diff
@@ -0,0 +1,397 @@
...
+	memberAddress := ids.ShortID{1}
+
+	tests := map[string]struct {
+		proposal                 *ExcludeMemberProposal
+		optionIndex              uint32
+		expectedProposalState    ProposalState
+		expectedOriginalProposal *ExcludeMemberProposal
+		expectedErr              error
+	}{
+		"Fail: option 2 out of bonds": {
```
Comment (knikos) [inline vms/platformvm/dac/camino_exclude_member_proposal_test.go]: small typo -> "..out of bo**u**nds"

**Review by knikos — COMMENTED** [review]
awesome test coverage and test case explanations. thank you!!

File: vms/platformvm/txs/executor/camino_helpers_test.go (hunk 1 above)
Comment (evlekht) [reply vms/platformvm/txs/executor/camino_helpers_test.go:739]: yep
for some later general clean-up / test-revision PR

File: vms/platformvm/txs/dac/camino_dac.go
Comment (evlekht) [reply vms/platformvm/txs/dac/camino_dac.go:1]: hmm, u'r right
I'll move it to executor

File: vms/platformvm/camino_helpers_test.go (hunk 3 above)
Comment (evlekht) [reply vms/platformvm/camino_helpers_test.go:118]: gh ci started to complain about it
I didn't investigate why, cause its not significant, though I was still surprised by it

File: vms/platformvm/dac/camino_proposal.go (hunk 4 above)
Comment (evlekht) [reply vms/platformvm/dac/camino_proposal.go]: I'd leave it as it is - its always better to have descriptive naming
e.g.
`getter.Invoke()`
vs
`getter.GetBondTxIDs`

even if we replace getter with full bondTxIDsGetter, imo, its still better to have more descriptive func naming to be 100% clear about what is happening

Comment (evlekht) [reply vms/platformvm/dac/camino_proposal.go:74]: I'd leave it as it is - its always better to have descriptive naming
e.g.
`getter.Invoke()`
vs
`getter.GetBondTxIDs`

even if we replace getter with full bondTxIDsGetter, imo, its still better to have more descriptive func naming to be 100% clear about what is happening

Comment (knikos) [reply vms/platformvm/dac/camino_proposal.go]: 100% with you on descriptive naming
Don't get me wrong i'm all about functions expressing their intentions.

There's always though the danger of going too verbose. One has to always strike a balance between verbosity and readability.

In the same sense that within a Storage object you wouldn't call method initializeStorage() but rather initialize() because you can infer the rest from the context, you wouldn't invoke a getter function by saying get(Getter) but rather something like invoke/call Getter because that's what you do. You don't get a getter , you call it.

In a different context where the argument of this function wouldn't be such a detailed/explanatory interface,
Then invoke would be too vague and thus not recommended.

Imagine a method called 

GetAverage(amount1,amount2)
And another called
GetAverageOfTwoAmounts(amount1,amount2)

The latter is much more descriptive but crosses the lines of necessary verbosity.

If the argument types provide sufficient context about the function's purpose, you can opt for a shorter, more straightforward name. This promotes readability and maintains a clean and concise codebase.

Comment (evlekht) [reply vms/platformvm/dac/camino_proposal.go]: upd: was further discussed outside of gh, that lead to renaming of other dac visitor interfaces and corresponding methods

**Review by knikos — DISMISSED** [review]

**Review by knikos — APPROVED** [review]
